apps/speed.c

```diff
@@ -561,52 +639,514 @@ static OPT_PAIR ecdh_choices[] = {
+
+static unsigned char *buf = NULL, *buf2 = NULL;
+static int j;
+static char *engine_id = NULL;
+
+
+#ifndef OPENSSL_NO_MD2
+static unsigned char md2[MD2_DIGEST_LENGTH];
+static int EVP_Digest_MD2_loop(void *args) {
+    int count;
+    for (count = 0; COND(c[D_MD2][j]); count++)
+        EVP_Digest(buf, (unsigned long)lengths[j], &(md2[0]), NULL,
+                EVP_md2(), NULL);
+    return count;
+}
+#endif
+
+#ifndef OPENSSL_NO_MDC2
+static unsigned char mdc2[MDC2_DIGEST_LENGTH];
+static int EVP_Digest_MDC2_loop(void *args) {
+    int count;
+    for (count = 0; COND(c[D_MDC2][j]); count++)
+        EVP_Digest(buf, (unsigned long)lengths[j], &(mdc2[0]), NULL,
+                EVP_mdc2(), NULL);
+    return count;
+}
+#endif
+
+#ifndef OPENSSL_NO_MD4
+static unsigned char md4[MD4_DIGEST_LENGTH];
+static int EVP_Digest_MD4_loop(void *args) {
+    int count;
+    for (count = 0; COND(c[D_MD4][j]); count++)
+        EVP_Digest(&(buf[0]), (unsigned long)lengths[j], &(md4[0]),
+                NULL, EVP_md4(), NULL);
+    return count;
+}
+#endif
+
+#ifndef OPENSSL_NO_MD5
+static unsigned char md5[MD5_DIGEST_LENGTH];
+static int MD5_loop(void *args) {
+    int count;
+    for (count = 0; COND(c[D_MD5][j]); count++)
+        MD5(buf, lengths[j], md5);
+    return count;
+}
+
+static unsigned char hmac[MD5_DIGEST_LENGTH];
+static HMAC_CTX *hctx = NULL;
+static int HMAC_loop(void *args) {
+    int count;
+    for (count = 0; COND(c[D_HMAC][j]); count++) {
+        HMAC_Init_ex(hctx, NULL, 0, NULL, NULL);
+        HMAC_Update(hctx, buf, lengths[j]);
+        HMAC_Final(hctx, &(hmac[0]), NULL);
+    }
+    return count;
+}
+#endif
+
+static unsigned char sha[SHA_DIGEST_LENGTH];
+static int SHA1_loop(void *args) {
+    int count;
+    for (count = 0; COND(c[D_SHA1][j]); count++)
+        SHA1(buf, lengths[j], sha);
+    return count;
+}
+
+static unsigned char sha256[SHA256_DIGEST_LENGTH];
+static int SHA256_loop(void *args) {
+    int count;
+    for (count = 0; COND(c[D_SHA256][j]); count++)
+        SHA256(buf, lengths[j], sha256);
+    return count;
+}
+
+static unsigned char sha512[SHA512_DIGEST_LENGTH];
+static int SHA512_loop(void *args) {
+    int count;
+    for (count = 0; COND(c[D_SHA512][j]); count++)
+        SHA512(buf, lengths[j], sha512);
+    return count;
+}
+
+#ifndef OPENSSL_NO_WHIRLPOOL
+static unsigned char whirlpool[WHIRLPOOL_DIGEST_LENGTH];
+static int WHIRLPOOL_loop(void *args) {
+    int count;
+    for (count = 0; COND(c[D_WHIRLPOOL][j]); count++)
+        WHIRLPOOL(buf, lengths[j], whirlpool);
+    return count;
+}
+#endif
+
+#ifndef OPENSSL_NO_RMD160
+static unsigned char rmd160[RIPEMD160_DIGEST_LENGTH];
+static int EVP_Digest_RMD160_loop(void *args) {
+    int count;
+    for (count = 0; COND(c[D_RMD160][j]); count++)
+        EVP_Digest(buf, (unsigned long)lengths[j], &(rmd160[0]), NULL,
+                EVP_ripemd160(), NULL);
+    return count;
+}
+#endif
+
+#ifndef OPENSSL_NO_RC4
+static RC4_KEY rc4_ks;
```

Comment:
I am moving the global buffers and the operation contexts inside the job.
Do you think that also the `RC4_KEY` and all the keys used by other operations (e.g. AES, DES, etc...) should not be shared across multiple jobs?